---
title: "Intro to Transcriptomics"
author: Nicole Black, Wade Boohar, Kayla Xu
date: 07/17/22
updated: 09/26/23
---

***Deliverables***
-Upload this R Notebook to your GitHub and submit the link to your Repo on BB.
-Include any graphs or figures created in this assignment in the folder with your R notebook with descriptive file names.

Since this is an optional partner activity, it is okay if your answers are the same as your partnerâ€™s as long as everyone understands it and could explain it in their own words if asked. Each person must individually push their code to Github. *At the top of your R Notebook, write the name of you and your partner(s) as a comment.*

***Complete the following coding activity and answer any following questions as comments in your R Notebook***

In Transcriptomics_Tutorial.Rmd, you learned how to manipulate the SummarizedExperiment data structure, turn it into more readable dataframes, complete basic analysis using those dataframes, and performed a differential expression analysis based on age. Before saving the rna_counts, rna_clinical, and rna_genes however, we masked out any sample that was not a tumor. In this assignment, you will need to reinstantiate the rna_se object and then perform differential expression analysis based on tumor status.

*Pre-Assignment*
Use knitr function to set your working directory to your analysis_data folder in qbio_490_name.
```{r}
 knitr::opts_knit$set(root.dir = normalizePath("/Users/ErikaLi/desktop/qbio/QBIO_490_erikali/analysis_data"))
```

Load all necessary packages and instantiate the rna_se object
```{r}
library(TCGAbiolinks)
library(SummarizedExperiment)


#GDCdownload(rna_query)
rna_query <- GDCquery(project ="TCGA-BRCA",
                      data.category = "Transcriptome Profiling",
                      data.type = "Gene Expression Quantification",
                      workflow.type = "STAR - Counts")

rna_se <- GDCprepare(rna_query)

```


*1*
Perform all data cleaning and label manipulation steps (hint: applying colnames and rownames to make the dataframes more readable) except for masking out the patient with an NA for age and the "Solid Tissue Normal" sample types.
```{r}
# colData
rna_clinical <- rna_se@colData

# Convert to a data frame
rna_clinical <- as.data.frame(rna_clinical)
colnames(rna_counts) <- rownames(rna_clinical)


#rowRanges
rna_genes <- rna_se@rowRanges@elementMetadata

# Set row names
rownames(rna_genes) <- colData(rna_se)$gene_id
rna_genes <- as.data.frame(rna_genes)
rownames(rna_genes) <- rna_genes$gene_id



```


*2*
Determine and define your experimental design for the differential expression analysis. Control for covariance of age, gender, OR vital_status. Manipulate those columns so that they are ready for differential expression analysis (hint: what kind of variables are they? what data type are they by default? do you need to handle unknown values?)
```{r}

# Convert categorical variables to factors
rna_clinical$vital_status <- as.factor(rna_clinical$vital_status)
rna_clinical$gender <- as.factor(rna_clinical$gender)

# Convert age to numeric
rna_clinical$age_at_diagnosis <- as.numeric(rna_clinical$age_at_diagnosis)
# Check for missing values in age
sum(is.na(rna_clinical$age_at_diagnosis))

#gender mask
age_na_mask <- !is.na(rna_clinical$age_at_diagnosis)
rna_clinical <- rna_clinical[age_na_mask, ]
rna_counts <- rna_counts[age_na_mask, ]

# Check for missing values in gender
sum(is.na(rna_clinical$gender))
# 1 NA value

#mask 1 missing value in gender in both data frames
gender_na_mask <- !is.na(rna_clinical$gender)
rna_clinical <- rna_clinical[gender_na_mask, ]
rna_counts <- rna_counts[gender_na_mask, ]


# Set up the design matrix
design_matrix <- model.matrix(~ age_at_diagnosis + gender + vital_status, data = rna_clinical)

```


*3*
Perform the differential expression analysis, create a volcano plot, and interpret what genes from each of the 6 sections means (assuming thresholds of -log10padj = 2 and log2FoldChange = 1)
```{r}

# Load necessary packages
install.packages("DESeq2")

library(DESeq2)
library(ggplot2)

# Extract counts
counts <- assay(rna_se)
sample_barcode <- rna_clinical$barcode

# since i masked one patient due to NA gender value in rna_clinical, I had to make sure the number of counts was equal to the number of rows in colData
counts_subset <- counts[, sample_barcode]

# Create DESeqDataSet
dds <- DESeqDataSetFromMatrix(countData = counts_subset, colData = rna_clinical, design = ~ age_at_diagnosis + gender + vital_status)

# Perform differential expression analysis
dds_obj <- DESeq(dds)

# Create results table
results <- results(dds_obj, contrast = c("gender", "male", "female"))

# Convert to data frame
results_df <- as.data.frame(results)


# Create volcano plot
volcano_plot <- ggplot(results_df, aes(x = log2FoldChange, y = -log10(pvalue))) +
  geom_point(aes(color = ifelse(results_df$padj < 0.01, "red", "black"))) +
  geom_hline(yintercept = 2, linetype = "dashed", color = "blue") +
  geom_vline(xintercept = c(-1, 1), linetype = "dashed", color = "blue") +
  labs(title = "Volcano Plot of Differential Expression",
       x = "Log2 Fold Change",
       y = "-Log10 Adjusted P-value") +
  theme_minimal()

# Save the plot
ggsave("/Users/ErikaLi/desktop/qbio/QBIO_490_erikali/analysis_data/volcano_plot.png", plot = volcano_plot)

# Display the volcano plot
print(volcano_plot)

```

top-right genes: genes that are highly upregulated and significant.

bottom-right genes: genes that are highly upregulated but not as significant.

top-left genes: genes that are highly downregulated and significant. 
bottom-left genes:genes that are highly downregulated but not as significant.

top-middle genes: genes that had significant results but were not differentially expressed. 

bottom-middle genes: genes that were neither differentially expressed nor significant. 

Save the picture of the volcano plot (using either ggsave() or right clicking and manually downloading the image and push this .Rmd and the image to GitHub)
